### PR TITLE
Misc ZonedDateTime and Duration changes to support relativeTo

### PR DESCRIPTION
**Pull request by ptomato**
See: #1171

### CODE REVIEW

**Comment by ptomato** [comment]
@justingrant - would you mind taking a close look at the tests in particular? It would be really helpful to have a second opinion that these are, in fact, the expected results.

**Comment by justingrant** [comment]
> @justingrant - would you mind taking a close look at the tests in particular? It would be really helpful to have a second opinion that these are, in fact, the expected results.

Yep, I'll review tomorrow after our meeting. Congrats for getting this done!

File: spec/duration.html [hunk 1]
```diff
@@ -1310,8 +1315,9 @@ <h1>RoundDuration ( _years_, _months_, _weeks_, _days_, _hours_, _minutes_, _sec
           1. Set _nanoseconds_ to 0.
         1. Else,
           1. Assert: _unit_ is *"nanoseconds"*.
-          1. Set _remainder_ to 0.
+          1. Set _remainder_ to _nanoseconds_.
           1. Set _nanoseconds_ to ? RoundNumberToIncrement(_nanoseconds_, _increment_, _roundingMode_).
+          1. Set _remainder_ to _remainder_ − _ nanoseconds_.
```
Comment (Ms2ger) [inline spec/duration.html:1320]:
```suggestion
          1. Set _remainder_ to _remainder_ − _nanoseconds_.
```

File: spec/duration.html [hunk 2]
```diff
@@ -1204,14 +1206,17 @@ <h1>MoveRelativeDate ( _calendar_, _relativeTo_, _duration_ )</h1>
...
         1. Let _years_, _months_, _weeks_, _days_, _hours_, _minutes_, _seconds_, _milliseconds_, _microseconds_, _nanoseconds_, and _increment_ each be the mathematical values of themselves.
-        1. If _unit_ is *"years"*, *"months"*, or *"weeks"*, then
-          1. If _relativeTo_ is not given, then
-            1. Throw a *RangeError* exception.
+        1. If _unit_ is *"years"*, *"months"*, or *"weeks"*, and _relativeTo_ is *undefined*, then
+          1. Throw a *RangeError* exception.
+        1. If _relativeTo_ is not *undefined*, then
+          1. If _relativeTo_ has an [[InitializedTemporalZonedDateTime]] internal slot, then
+            1. Let _instant_ be ? CreateTemporalInstant(_relativeTo_[[Nanoseconds]]).
+            1. Let _relativeTo_ be ? GetTemporalDateTimeFor(_relativeTo_.[[TimeZone]], _instant_, _relativeTo_.[[Calendar]]).
```
Comment (Ms2ger) [inline spec/duration.html:1214]:
```suggestion
            1. Set _relativeTo_ to ? GetTemporalDateTimeFor(_relativeTo_.[[TimeZone]], _instant_, _relativeTo_.[[Calendar]]).
```

File: spec/duration.html [hunk 3]
```diff
@@ -1328,6 +1334,63 @@ <h1>RoundDuration ( _years_, _months_, _weeks_, _days_, _hours_, _minutes_, _sec
...
+            [[Milliseconds]]: _milliseconds_,
+            [[Microseconds]]: _microseconds_,
+            [[Nanoseconds]]: _nanoseconds_
+            }.
+        1. Let _timeRemainderNs_ be ? TotalDurationNanoseconds(0, _hours_, _minutes_, _seconds_, _milliseconds_, _microseconds_, _nanoseconds_, 0).
+        1. Let _direction_ be ! Sign(_timeRemainderNs_).
+        1. Let _dayStart_ be ? AddZonedDateTime(_relativeTo_.[[Nanoseconds]], _relativeTo_.[[TimeZone]], _relativeTo_.[[Calendar]], _years_, _months_, _weeks_, _days_, 0, 0, 0, 0, 0, 0, *"constrain"*).
+        1. Let _dayEnd_ be ? AddZonedDateTime(_dayStart_, _relativeTo_.[[TimeZone]], _relativeTo_.[[Calendar]], 0, 0, 0, _direction_, 0, 0, 0, 0, 0, 0, *"constrain"*).
+        1. Let _dayLengthNs_ be _dayEnd_ − _dayStart_.
+        1. If (_timeRemainderNs_ − _dayLengthNs_) × _direction_ < 0, then
```
Comment (Ms2ger) [inline spec/duration.html:1363]:
```suggestion
        1. If (_timeRemainderNs_ − _dayLengthNs_) × _direction_ &lt; 0, then
```

**Review by Ms2ger — APPROVED** [review]
I'll approve now, but hold off on merging pending @justingrant's review.

**Comment by justingrant** [comment]
@ptomato and I discussed. Decisions:
 
* `Duration.prototype.add` with `relativeTo` will be implemented so that `result1` and `result2` below are always equal. (I'm omitting the conversion of `relativeTo` inputs to either PlainDateTime or ZonedDateTime, which shouldn't affect the results.)
```js
result1 = duration.add(other, options);
result2 = options.relativeTo.until(
  options.relativeTo.add(duration).add(other), 
  options
);
```
* tests will include cases with three variations: 1) both `this` and `other` are positive; 2) both are negative; 3) one is positive, one is negative
* `Duration.prototype.round` will work similarly to `add`, except only with one `add` operation. Like this: 
```js
result1 = duration.round(options);
result2 = options.relativeTo.until(
  options.relativeTo.add(duration), 
  options
);
```
* Rounding tests should include negative and positive durations where the duration includes a DST transition

**Comment by ptomato** [comment]
Indeed, I think the realization for me today was that the order matters in `Duration.from({ hours: 24 }).add({ days: 1, relativeTo })` — missing that is the reason why the tests I originally wrote were not correct.

**Comment by ptomato** [comment]
I reviewed the tests with @justingrant on a call today and I've verified them now with the above pseudocode.

**Comment by justingrant** [comment]
Why are the tests still skipped?

**Comment by ptomato** [comment]
This PR doesn't contain the fixes that make them work ... I'm still working on those.

**Comment by justingrant** [comment]
Oh, got it. I thought this was going to include those changes. Sorry, with so many different PRs zipping around, easy to get lost. ;-(

**Pull request merged**